src/Workspaces/MSBuild/BuildHost/MSBuild/ProjectFile/ProjectInstanceReader.cs

```diff
@@ -0,0 +1,256 @@
+            DefaultNamespace = defaultNamespace,
+            TargetFramework = targetFramework,
+            TargetFrameworkIdentifier = targetFrameworkIdentifier,
+            TargetFrameworkVersion = targetFrameworkVersion,
+            ProjectAssetsFilePath = projectAssetsFilePath,
+            CommandLineArgs = commandLineArgs,
+            Documents = docs,
+            AdditionalDocuments = additionalDocs,
+            AnalyzerConfigDocuments = analyzerConfigDocs,
+            ProjectReferences = [.. _projectInstance.GetProjectReferences()],
+            PackageReferences = packageReferences,
+            ProjectCapabilities = projectCapabilities,
+            ContentFilePaths = contentFileInfo,
+            FileGlobs = fileGlobs
+        };
+
+        static FileGlobs GetFileGlobs(MSB.Evaluation.GlobResult g)
+        {
+            return new FileGlobs(
+                Includes: [.. g.IncludeGlobs.Select(PathUtilities.ExpandAbsolutePathWithRelativeParts)],
+                Excludes: [.. g.Excludes.Select(PathUtilities.ExpandAbsolutePathWithRelativeParts)],
+                Removes: [.. g.Removes.Select(PathUtilities.ExpandAbsolutePathWithRelativeParts)]);
+        }
+    }
+
+    private static ImmutableArray<string> GetContentFiles(MSB.Execution.ProjectInstance project)
+    {
+        var contentFiles = project
+            .GetItems(ItemNames.Content)
+            .SelectAsArray(item => item.GetMetadataValue(MetadataNames.FullPath));
+        return contentFiles;
+    }
+
+    private ImmutableArray<string> GetCommandLineArgs(MSB.Execution.ProjectInstance project)
+    {
+        var commandLineArgs = _commandLineProvider.GetCompilerCommandLineArgs(project)
+            .SelectAsArray(item => item.ItemSpec);
+
+        if (commandLineArgs.Length == 0)
+        {
+            // We didn't get any command-line args, which likely means that the build
+            // was not successful. In that case, try to read the command-line args from
+            // the ProjectInstance that we have. This is a best effort to provide something
+            // meaningful for the user, though it will likely be incomplete.
+            commandLineArgs = _commandLineProvider.ReadCommandLineArgs(project);
+        }
+
+        return commandLineArgs;
+    }
+
+    private static bool IsNotTemporaryGeneratedFile(MSB.Framework.ITaskItem item)
+        => !Path.GetFileName(item.ItemSpec).StartsWith("TemporaryGeneratedFile_", StringComparison.Ordinal);
+
+    private DocumentFileInfo MakeDocumentFileInfo(MSB.Framework.ITaskItem documentItem)
+    {
+        var filePath = GetDocumentFilePath(documentItem);
+        var logicalPath = GetDocumentLogicalPath(documentItem, _projectDirectory);
+        var isLinked = IsDocumentLinked(documentItem);
+
+        var folders = GetRelativeFolders(documentItem);
+        return new DocumentFileInfo(filePath, logicalPath, isLinked, isGenerated: false, folders);
```

Comment:
Re `isGenerated: false`

This used to check if the Compile item is present in the project file vs project instance. It considered items that were only present in the instance as generated. This is rather useless heuristic, since it's entirely dependent on the implementation of the targets that generate source files whether they pre-define the Compile items during evaluation or add them only during DTB.

It doesn't look like any of our MSBuildWorkspace usages rely on this value anyways.

Reply:
https://github.com/dotnet/roslyn/issues/81755